cql-pytest: translate Cassandra's tests for SELECT operations

nyh opened the pull request:
This is a translation of Cassandra's CQL unit test source file
validation/operations/SelectTest.java into our our cql-pytest framework.

This large test file includes 78 tests for various types of SELECT
operations. Four additional tests require UDF in Java syntax,
and were skipped.

All 78 tests pass on Cassandra. 25 of the tests fail on Scylla
reproducing 3 already known Scylla issues and 8 previously-unknown
issues:

Previously known issues:

Refs https://github.com/scylladb/scylla/issues/2962:  Collection column indexing
Refs https://github.com/scylladb/scylla/issues/4244:  Add support for mixing token, multi- and single-column restrictions
Refs https://github.com/scylladb/scylla/issues/8627:  Cleanly reject updates with indexed values where value > 64k

Newly-discovered issues:

Refs https://github.com/scylladb/scylla/issues/10354: SELECT DISTINCT should allow filter on static columns, not just partition keys
Refs https://github.com/scylladb/scylla/issues/10357: Spurious static row returned from query with filtering, despite not matching filter
Refs https://github.com/scylladb/scylla/issues/10358: Comparison with UNSET_VALUE should produce an error
Refs https://github.com/scylladb/scylla/issues/10359: "CONTAINS NULL" and "CONTAINS KEY NULL" restrictions should match nothing
Refs https://github.com/scylladb/scylla/issues/10361: Null or UNSET_VALUE subscript should generate an invalid request error
Refs https://github.com/scylladb/scylla/issues/10366: Enforce Key-length limits during SELECT
Refs https://github.com/scylladb/scylla/issues/10443: SELECT with IN and ORDER BY orders rows per partition instead of for the entire response
Refs https://github.com/scylladb/scylla/issues/10448: The CQL token() function should validate its parameters

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/189/

nyh commented:
It turns out that one of the tests, testSelectWithAlias(), sets the ttl of a cell to 10, and then tries to read it and expects to see 10 or 9 but... in our **debug** build and ridiculously overloaded setup it gets (see https://jenkins.scylladb.com/job/releng/job/Scylla-CI/189/testReport/cassandra_tests.validation.operations/select_test/Unit_Tests___testSelectWithAlias/) 3, i.e., a couple of requests took 7 seconds.

I'll remove that part of the test.

nyh commented:
Forced-push a version that:
1. Changed the TTL-reading tests to allow the test to take 30 seconds instead of just 1. In https://jenkins.scylladb.com/job/releng/job/Scylla-CI/189/ in the debug build, the test could take 7 seconds. Note that the test doesn't actually take 30 or even 1 second (in a normal build on a normal machine, it takes a small fraction of a second).
2. Removed the first patch, which was already committed. I'll edit the cover letter accordingly as this is now a one-patch PR.

nyh commented:
Unfortunately there are new failures in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/194/, I'll debug them.

By the way, @benipeled do you know why the CI above is reporting "unit tests completed successfully" and "waiting for sanity tests status to be reported", while in fact the unit tests completed **unsuccessfully**, and I don't know what is "sanity tests"?

nyh commented:
@benipeled the test in this PR failed because it needs a fix in "next", not yet in "master", to work. Could it be that the CI is merging this PR into master instead of into next? If this is the case, it's wrong (I think).

Also, now that the test failed because it used the wrong base branch - how do I cause the test to run again when "master" gets updated?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/194/

nyh commented:
@benipeled the text above now says "CI state FAILURE" in a message, but "All checks have passed" in the green frame. This is inconsistent :-( I'll rebase and re-push the branch to have it retested.

nyh commented:
Rebased and forced-pushed to get the CI to test this patch again (after the prerequisites reached master). When the CI finally approves, please review and consider merging.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/222/
